test/src/test/integration/link_git_protocol.rs

```diff
@@ -6,91 +6,104 @@
 use std::{
     collections::BTreeSet,
     io,
-    path::Path,
+    path::{Path, PathBuf},
     sync::{atomic::AtomicBool, Arc},
 };
 
 use bstr::ByteSlice as _;
 use futures::{AsyncReadExt as _, TryFutureExt as _};
+use git::{
+    prelude::*,
+    refs::transaction::{Change, PreviousValue, RefEdit},
+};
+use git_repository as git;
 use link_git_protocol::{fetch, ls, packwriter, upload_pack, ObjectId, PackWriter, Ref};
 use tempfile::{tempdir, TempDir};
 
 fn upstream() -> TempDir {
     let tmp = tempdir().unwrap();
 
-    let repo = git2::Repository::init_bare(&tmp).unwrap();
-    let auth = git2::Signature::now("apollo", "apollo@cree.de").unwrap();
-
-    let tree = {
-        let empty = repo.treebuilder(None).unwrap();
-        let oid = empty.write().unwrap();
-        repo.find_tree(oid).unwrap()
-    };
-    let base = {
-        let oid = repo
-            .commit(
-                Some("refs/namespaces/foo/refs/heads/main"),
-                &auth,
-                &auth,
-                "initial",
-                &tree,
-                &[],
-            )
-            .unwrap();
-        repo.find_commit(oid).unwrap()
+    let repo = git::init_bare(&tmp).unwrap().into_easy();
+    let auth = git::actor::Signature::now_local_or_utc("apollo", "apollo@cree.de");
+
+    let empty_tree_id = repo
+        .write_object(&git::objs::Tree::empty())
```

Comment:
Now one can write object refs as well as owned objects, and there is a trait for that so calls look as good as they can.